compiler/noirc_evaluator/src/ssa/opt/inlining.rs

```diff
@@ -46,54 +46,80 @@ impl Ssa {
-                        }
-                        RuntimeType::Brillig(_) => {
-                            // If the called function is brillig, we inline only if it's into brillig and the function is not recursive
-                            ssa.functions[entry_point].runtime().is_brillig()
-                                && !inline_sources.contains(&called_func_id)
-                        }
-                    }
-                };
+        let inline_targets =
+            inline_infos.iter().filter_map(|(id, info)| info.is_inline_target().then_some(*id));
 
+        // NOTE: Functions are processed independently of each other, with the final mapping replacing the original,
+        // instead of inlining the "leaf" functions, moving up towards the entry point.
+        self.functions = btree_map(inline_targets, |entry_point| {
+            let function = &self.functions[&entry_point];
             let new_function =
-                InlineContext::new(&self, *entry_point).inline_all(&self, &should_inline_call);
-            (*entry_point, new_function)
+                function.inlined(&self, inline_no_predicates_functions, inline_infos);
+            (entry_point, new_function)
         });
         self
     }
 }
 
+impl Function {
+    /// Create a new function which has the functions called by this one inlined into its body.
+    pub(super) fn inlined(
+        &self,
+        ssa: &Ssa,
+        inline_no_predicates_functions: bool,
+        inline_infos: &InlineInfos,
+    ) -> Function {
+        let should_inline_call =
+            |_context: &PerFunctionContext, ssa: &Ssa, called_func_id: FunctionId| -> bool {
+                // Do not inline self-recursive functions on the top level.
+                // Inlining a self-recursive function works when there is something to inline into
+                // by importing all the recursive blocks, but for the entry function there is no wrapper.
+                if called_func_id == self.id() {
+                    return false;
+                }
+                let callee = &ssa.functions[&called_func_id];
+
+                match callee.runtime() {
+                    RuntimeType::Acir(inline_type) => {
+                        // If the called function is acir, we inline if it's not an entry point
+
+                        // If we have not already finished the flattening pass, functions marked
+                        // to not have predicates should be preserved.
+                        let preserve_function =
+                            !inline_no_predicates_functions && callee.is_no_predicates();
+
+                        !inline_type.is_entry_point() && !preserve_function
+                    }
+                    RuntimeType::Brillig(_) => {
+                        // If the called function is brillig, we inline only if it's into brillig and the function is not recursive and not too costly.
```

Comment:
Can you improve this comment? Not sure what "we inline only if it's into brillig" means. Do you mean we inline only if its caller is brillig?

Reply:
Yes; I'll come up with better wording 👍